src/rc_string.c

```diff
@@ -7,4 +7,168 @@
+            rc_string_internal->storage_type = STRING_STORAGE_TYPE_COPIED;
+
+            /* Codes_SRS_RC_STRING_01_004: [ rc_string_create shall copy the string memory (including the NULL terminator). ]*/
+            (void)memcpy((void*)rc_string_internal->rc_string.string, string, string_length_with_terminator);
+
+            /* Codes_SRS_RC_STRING_01_005: [ rc_string_create shall succeed and return a non-NULL handle. ]*/
+            THANDLE_MOVE(RC_STRING)(&result, &temp_result);
+        }
+    }
+
+    return result;
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, THANDLE(RC_STRING), rc_string_create_with_move_memory, const char*, string)
+{
+    THANDLE(RC_STRING) result = NULL;
+
+    if (string == NULL)
+    {
+        /* Codes_SRS_RC_STRING_01_007: [ If string is NULL, rc_string_create_with_move_memory shall fail and return NULL. ]*/
+        LogError("Invalid arguments: const char*, string=%s", MU_P_OR_NULL(string));
+    }
+    else
+    {
+        /* Codes_SRS_RC_STRING_01_008: [ Otherwise, rc_string_create_with_move_memory shall allocate memory for the THANDLE(RC_STRING). ]*/
+        THANDLE(RC_STRING) temp_result = THANDLE_MALLOC_WITH_EXTRA_SIZE(RC_STRING)(rc_string_dispose, sizeof(RC_STRING_INTERNAL) - sizeof(RC_STRING));
+        if (temp_result == NULL)
+        {
+            /* Codes_SRS_RC_STRING_01_011: [ If any error occurs, rc_string_create_with_move_memory shall fail and return NULL. ]*/
+            LogError("THANDLE_MALLOC_WITH_EXTRA_SIZE(RC_STRING) failed");
+        }
+        else
+        {
+            RC_STRING_INTERNAL* rc_string_internal = RC_STRING_INTERNAL_FROM_RC_STRING(THANDLE_GET_T(RC_STRING)(temp_result));
+
+            /* Codes_SRS_RC_STRING_01_009: [ rc_string_create_with_move_memory shall associate string with the new handle. ]*/
+            rc_string_internal->rc_string.string = string;
+            rc_string_internal->storage_type = STRING_STORAGE_TYPE_MOVED;
+
+            /* Codes_SRS_RC_STRING_01_010: [ rc_string_create_with_move_memory shall succeed and return a non-NULL handle. ]*/
+            THANDLE_MOVE(RC_STRING)(&result, &temp_result);
+        }
+    }
+
+    return result;
+}
+
+IMPLEMENT_MOCKABLE_FUNCTION(, THANDLE(RC_STRING), rc_string_create_with_custom_free, const char*, string, RC_STRING_FREE_FUNC, free_func, void*, free_func_context)
+{
+    THANDLE(RC_STRING) result = NULL;
+
+    /* Codes_SRS_RC_STRING_01_014: [ free_func_context shall be allowed to be NULL. ]*/
+
+    if (
+        /* Codes_SRS_RC_STRING_01_012: [ If string is NULL, rc_string_create_with_custom_free shall fail and return NULL. ]*/
+        (string == NULL) ||
+        /* Codes_SRS_RC_STRING_01_013: [ If free_func is NULL, rc_string_create_with_custom_free shall fail and return NULL. ]*/
+        (free_func == NULL)
+        )
+    {
+        LogError("Invalid arguments: const char* string=%s, RC_STRING_FREE_FUNC free_func=%p, void* free_func_context=%p", MU_P_OR_NULL(string), free_func, free_func_context);
```

Comment:
>* s [](start = 47, length = 3)

missing "," to be consistent with the other LogErrors in the same place. Or ever better - change the other LogErrors not to have ","

Reply:
Removed , in other places

---
In reply to: [460259212](https://github.com/Azure/azure-c-util/pull/52#discussion_r460259212) [](ancestors = 460259212)